File: common/src/main/java/com/imcys/bilibilias/common/utils/BiliAppSigner.java
package com.imcys.bilibilias.common.utils;


import java.io.UnsupportedEncodingException;
import java.util.Map;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.net.URLEncoder;
import java.util.TreeMap;
public class BiliAppSigner {
    public static final String APP_KEY = "27eb53fc9058f8c3";
    public static final String APP_SEC = "c2ed53a74eeefe3cf99fbd01d8c9c375";

    public static String appSign(Map<String, String> params) {
        // 为请求参数进行 APP 签名
        params.put("appkey", APP_KEY);
        // 按照 key 重排参数
        Map<String, String> sortedParams = new TreeMap<>(params);
        // 序列化参数
        StringBuilder queryBuilder = new StringBuilder();
        for (Map.Entry<String, String> entry : sortedParams.entrySet()) {
            if (queryBuilder.length() > 0) {
                queryBuilder.append('&');
            }

            try {
                queryBuilder
                        .append(URLEncoder.encode(entry.getKey(), "UTF-8"))
                        .append('=')
                        .append(URLEncoder.encode(entry.getValue(), "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                return null;
            }
        }
        return generateMD5(queryBuilder .append(APP_SEC).toString());
    }

    private static String generateMD5(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(input.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        return null;
    }

}

File: core/network/src/main/java/com/imcys/bilibilias/network/utils/BiliAppSigner.kt
package com.imcys.bilibilias.network.utils

import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TreeMap
import kotlin.random.Random


object BiliAppSigner {
    const val APP_KEY: String = "4409e2ce8ffd12b8"
    const val APP_SEC: String = "59b43e04ad6965f34319062b478f83dd"

    private val CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_0123456789".toCharArray()

    fun appSign(params: MutableMap<String, String>): String? {
        // 为请求参数进行 APP 签名
        params.put("appkey", APP_KEY)
        // 按照 key 重排参数
        val sortedParams: MutableMap<String, String> = TreeMap<String, String>(params)
        // 序列化参数
        val queryBuilder = StringBuilder()
        for (entry in sortedParams.entries) {
            if (queryBuilder.isNotEmpty()) {
                queryBuilder.append('&')
            }

            try {
                queryBuilder
                    .append(URLEncoder.encode(entry.key, "UTF-8"))
                    .append('=')
                    .append(URLEncoder.encode(entry.value, "UTF-8"))
            } catch (e: UnsupportedEncodingException) {
                return null
            }
        }
        return generateMD5(queryBuilder.append(APP_SEC).toString())
    }

    private fun generateMD5(input: String): String? {
        try {
            val md = MessageDigest.getInstance("MD5")
            val digest = md.digest(input.toByteArray())
            val sb = StringBuilder()
            for (b in digest) {
                sb.append(String.format("%02x", b))
            }
            return sb.toString()
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
        }
        return null
    }

    private fun randomString(len: Int, rnd: Random = Random.Default): String =
        buildString(len) { repeat(len) { append(CHARS[rnd.nextInt(CHARS.size)]) } }


    val biliTvDeviceInfo by lazy {
        val now = Date()
        val sdf = SimpleDateFormat("yyyyMMddHHmmssSSS", Locale.CHINA)
        val deviceId = randomString(20)
        val buvid = randomString(37)
        val fingerprint = sdf.format(now) + randomString(45)

        val p = mutableMapOf<String, String>()
        p["bili_local_id"] = deviceId
        p["build"] = "102801"
        p["buvid"] = buvid
        p["channel"] = "master"
        p["device"] = "OnePlus"
        p["device_id"] = deviceId
        p["device_name"] = "OnePlus7TPro"
        p["device_platform"] = "Android10OnePlusHD1910"
        p["fingerprint"] = fingerprint
        p["guid"] = buvid
        p["local_fingerprint"] = fingerprint
        p["local_id"] = buvid
        p["mobi_app"] = "android_tv_yst"
        p["networkstate"] = "wifi"
        p["platform"] = "android"
        p["sys_ver"] = "29"
        p
    }
}